### ISSUE TITLE
Rename tardigrade->storj and clarify s3/native backend use cases

### ISSUE THREAD

**Issue by elek** [rclone/rclone#5616]
#### The associated forum post URL from `https://forum.rclone.org`

https://forum.rclone.org/t/backward-compatibility-rules/26569/12

#### What is your current rclone version (output from `rclone version`)?

```
clone v1.57.0-beta.5661.1dd360a01
- os/version: arch (64 bit)
- os/kernel: 5.14.2-arch1-2 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.17
- go/linking: dynamic
- go/tags: none
```

#### What problem are you are trying to solve?

Tardigrade was renamed to Storj (or Storj DCS) [recently](https://storj.io/blog/introducing-storj-dcs-decentralized-cloud-storage-for-developers).

To avoid confusion it seems to be better to do the rename on rclone side too.

During the forum discussion, it was suggested to explain better the advantages and disadvantages of using s3/native storj backend as storj also support s3 interface via `gateway-mt`.

S3 gateway can be useful for edge clients as less CPU/bandwidth is required, however, these CPU/bandwidth are currently provided by Storj Labs Inc (for free as of today). 

Native tardigrade/storj backend uses slightly more CPU and bandwidth (for upload) but it uses the decentralized cloud of storage nodes and it's more secure (as the secret credential is not shared with central service).

@olefrost suggest the following action points in the forums:

 * [x] Make a GitHub issue to discuss the overall approach and plan
 * [x] Confirm full Storj S3 compatibility by running the automated S3 backend test against an S3 TestDrive having Storj as an endpoint.
 * [ ] Make a pull request to Add Storj to the list of S3 providers on [this page](https://github.com/rclone/rclone/blob/master/docs/content/s3.md)
 * [x] Make a performance comparison of the S3 and tardigrade backends to collect information on the expected resource usage and speed on representative use cases using identical equipment.
 * [ ] Make a pull request to update the tardigrade (and S3) backend documentation with this information.
  * [ ] Make a pull request to rename the tardigrade backend to Storj - or to depreciate it in favor of Storj’s own client (https://www.storj.io/integrations/uplink-cli). The approach is to be based on the results and agreements obtained in the previous steps.
 
#### How do you think rclone should be changed to solve that?

The rename can be done with a backward-compatible way (POC is in #5613) with keeping the tardigrade backend as a pointer to the storj.

Documentation should be added to clarify s3 compatibility...

**Comment by ncw** [comment]
@calebcase as you wrote the original tardigrade backend, do you have any thoughts on whether we should rename it from `tardigrade` to `storj`?

**Comment by elek** [comment]
About the second point (`Confirm full Storj S3 compatibility`)

```
cd backend/s3
go test -v -remote storjs3: -fast-list -json
```

76 test passed, 35 failed

https://gateway.ipfs.io/ipfs/QmWZN1xLYP2BH97UC7xyXWiUDWueV2DDLG3T555cB6Dfzr

**Comment by elek** [comment]
>  Make a performance comparison of the S3 and tardigrade backends to collect information on the expected resource usage and speed on representative use cases using identical equipment.

I did a quick test from a server from a EU datacenter:

```
VENDOR
	Manufacturer: FUJITSU
	Product Name:  
MEMORY
	Size: 16384 MB
	Size: 16384 MB
CPUs
	Version: Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
	Max Speed: 8300 MHz
	Core Enabled: 4
	Thread Count: 8
DISKS
     NAME     SIZE TYPE  FSTYPE            MOUNTPOINT VENDOR   MODEL
     sda      3.7T disk                               ATA      TOSHIBA_MG04ACA400EY
     ├─sda1   3.7T part  linux_raid_member                     
     │ └─md0  3.7T raid1 ext4              /                   
     └─sda2     1M part                                        
     sdb      3.7T disk                               ATA      TOSHIBA_MG04ACA400EY
     ├─sdb1   3.7T part  linux_raid_member                     
     │ └─md0  3.7T raid1 ext4              /                   
     └─sdb2     1M part
```

I downloaded a 10Gb files multiple times.

Using tardigrade protocol, it took ~3.5 minutes: 

```
elek@os:~/prog/rclone-v1.56.1-linux-amd64$ time rclone copy -P stdom://uploadpath/10GB_test.img 10gx
Transferred:   	      10Gi / 10 GiByte, 100%, 47.150 MiByte/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:      3m36.9s
```

Using us gateway the time was almost the double and s3 protocol:

```
elek@os:~/prog/rclone-v1.56.1-linux-amd64$ time rclone copy -P stdoms3://uploadpath/10GB_test.img 10gx3
Transferred:   	      10Gi / 10 GiByte, 100%, 6.425 MiByte/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:      7m57.9s

real	7m57.984s
user	1m21.738s
sys	1m32.458s

elek@os:~/prog/rclone-v1.56.1-linux-amd64$ time rclone copy -P stdoms3://uploadpath/10GB_test.img 10gx4
Transferred:   	      10Gi / 10 GiByte, 100%, 5.465 MiByte/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:       8m6.3s

real	8m6.398s
user	1m32.435s
sys	1m44.404s
```

(The eu gateway is way faster, again).

I think it shows that we need both the uplink based tardigrade/storj and s3 based backend. 

Beyond the security/privacy factor (only the tardigrade/storj has full e2e encryption) the performance also can be better in some cases (download from server) using the native client.

**Comment by olefrost** [comment]
Nice testing - thanks!

I haven’t done backend testing before and don’t know if the 35 failed integration tests are an OK result, or an indication of issues in rclone and/or the StorJ S3 interface. Let’s wait and hear what the backend experts have to say.

Your performance test however clearly shows that the Tardigrade/Storj backend has a justification in scenarios with big servers connected directly to the internet backbone.

The performance test also shows that the S3 gateway servers can deliver sustained downloads up to app. 21 MB/s (10000/8/60).
This indicates, that rclone users with free internet download capacity below app. 300 Mbit/s will experience the best download speeds from the S3 protocol.

It would be useful to also have an upload test to make a similar estimate for uploads. This may well be the main limiter of two reasons: 1) Uploads have the highest Storj bandwidth multiplier 2) Many (consumer) internet connections have lower upload than download speed - e.g. 300/30Mbps.

**Comment by elek** [comment]
> It would be useful to also have an upload test to make a similar estimate for uploads. 

Yeah, good question. Here is a simple try (not a proper performance test, I just tried out from the same server):

```
elek@os:~/prog/rclone-v1.56.1-linux-amd64$ ./rclone copy -P 10g storjtest:bucket1/test1
                       
Transferred:        9.766Gi / 9.766 GiByte, 100%, 14.356 MiByte/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:      9m57.0s
```

```
elek@os:~/prog/rclone-v1.56.1-linux-amd64$ ./rclone copy -P 10g storjs3test:bucket1/test2
Transferred:        9.766Gi / 9.766 GiByte, 100%, 7.167 MiByte/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:     21m29.5s
```

So s3 is 2x slower. I guess the pure storj uplink cli is even faster as It can do parallel upload (means, uploading one huge local file parallel). I didn't find anything similar in rclone. As far as I see only file level parallelism is supported: different files can be uploaded parallel but not different parts of the same file.

**Comment by ncw** [comment]
@elek wrote
> About the second point (`Confirm full Storj S3 compatibility`)
> 
> ```
> cd backend/s3
> go test -v -remote storjs3: -fast-list -json
> ```
> 
> 76 test passed, 35 failed
> 
> https://gateway.ipfs.io/ipfs/QmWZN1xLYP2BH97UC7xyXWiUDWueV2DDLG3T555cB6Dfzr

It looks like their implementation doesn't support Copy or some punctuation in file names which explains a lot of the errors.

This failure needs looking into though...

```
TestIntegration/FsMkdir/FsPutChunked/10M/104857604.61s

=== RUN   TestIntegration/FsMkdir/FsPutChunked/10M/10485760
    fstest.go:126: 
        	Error Trace:	fstest.go:126
        	            				fstest.go:131
        	            				fstest.go:150
        	            				fstests.go:245
        	Error:      	Should be true
        	Test:       	TestIntegration/FsMkdir/FsPutChunked/10M/10485760
        	Messages:   	chunked-10M-10M.bin: Modification time difference too big |-180916h18m10.500000001s| > 1ns (want 2001-02-03 04:05:06.499999999 +0000 UTC vs got 2021-09-24 08:23:17 +0000 UTC) (precision 1ns)
    fstest.go:126: 
        	Error Trace:	fstest.go:126
        	            				fstest.go:131
        	            				fstest.go:150
        	            				fstests.go:249
        	Error:      	Should be true
        	Test:       	TestIntegration/FsMkdir/FsPutChunked/10M/10485760
        	Messages:   	chunked-10M-10M.bin: Modification time difference too big |-180916h18m10.500000001s| > 1ns (want 2001-02-03 04:05:06.499999999 +0000 UTC vs got 2021-09-24 08:23:17 +0000 UTC) (precision 1ns)
                --- FAIL: TestIntegration/FsMkdir/FsPutChunked/10M/10485760 (4.61s)
```

That looks like the metadata didn't get uploaded for that chunked file. The fact that it worked for other chunked files means to me that this is probably a Storj bug...

If they don't support Copy then the s3 backend needs to know about it so it can report that to the rest of rclone. No copy means no renaming of files or setting modification times - it is quite a big loss.

**Comment by olefrost** [comment]
> Here is a simple try (not a proper performance test, I just tried out from the same server)

Thanks, that’s fine - and enough for me to get a good picture of the performance differences.

The S3 gateway maintains a sustained upload of app. 8 MiByte/s (10000/21.5/60). This indicates that rclone users with free internet upload capacity below app. 250 Mbit/s will experience the best upload speeds from the S3 protocol.

Based on this I would say that the S3 protocol is the best and simplest way to interface to StorJ for the ordinary rclone user (the default so to speak). The Tardigrade/StorJ protocol is interface for pro users having the equipment, internet connection (above 300/300 Mbit/s) and technical skills to configure and troubleshoot such an environment, e.g. increase the maximum number of open files/tcp connections on the system (requiring root/admin access and skills) - and of cause anybody with an interest in the distributed security model of Storj.

I think we should somehow make this clear in the the docs, to guide users to the right solution - if the S3 functionality is satisfactory - and we all agree.

On a curios side note: Do you know how and where the directory entries (the mapping from a bucket name to the individual pieces of a file) is stored? Is it central or distributed? (a link is fine)

> I guess the pure storj uplink cli is even faster as It can do parallel upload (means, uploading one huge local file parallel). I didn't find anything similar in rclone. As far as I see only file level parallelism is supported: different files can be uploaded parallel but not different parts of the same file.

The rclone S3 interface supports chunked upload of file parts and would do it in your test case - if I understand the docs correctly. See: https://rclone.org/s3/#multipart-uploads , https://rclone.org/s3/#s3-upload-cutoff and https://rclone.org/s3/#s3-chunk-size 

> It looks like their implementation doesn't support Copy or some punctuation in file names which explains a lot of the errors.

How does the test result compare to the current Tardigrade/Storj backend? 
(I looked a bit around on GitHub and the integration server, but didn’t find it)

**Comment by elek** [comment]
> On a curios side note: Do you know how and where the directory entries (the mapping from a bucket name to the individual pieces of a file) is stored? Is it central or distributed? (a link is fine)

TLDR; It's more like centralish.

They are stored encrypted in the satellites (this is where your encryption key is local-only if you use tardigrade backend). Storage nodes can select the satellites to be joined to.

Today there are three main satellites and all of them are maintained by Storj Inc, but the architecture intentionally supports independent satellites, too (the question of more, 3rd party satellites usually pops up in community, last time discussed on [Fireside chat](https://forum.storj.io/t/node-operator-fireside-discussion/15297/12).

> The rclone S3 interface supports chunked upload of file parts and would do it in your test case - if I understand the docs correctly

Yes, but (fix me if I am wrong) it uses memory. As there is no access to the original file which can be read with random access, it sequentially reads the first 4 chunks to the memory and uploads parallel to the backend.

This is not a problem with the standard S3 backend, where the default chunk size is 5MB. But Storj would prefer higher chunk size (let's say 64 MB). In this case, the memory consumption would be more significant (4 * 64MB +...)

This is because rclone is general and supports copy from any backend to any other backend.

Uplink cli for Storj uploads only from local files. Therefore, it's possible to read the file from 0,1,2,3... segment offset at the same time without using buffer segments. 
 
 > How does the test result compare to the current Tardigrade/Storj backend?
 
 Good question. Just re-executed the test suite in tardigrade backend and everything will be passed: 98 passed, 17 skipped, 0 failed
 
> I think we should somehow make this clear in the the docs, to guide users to the right solution - if the S3 functionality is satisfactory - and we all agree.

Sounds good to me, I created a #5646 where the documentation is copied from Storj both to the Tardigrade and s3 backends Please let me know there if you would explain it in more details (I think the bandwidth amplification is clearly explained)

**Comment by elek** [comment]
BTW, I am planning to check the S3 failures deeper, thanks to point to this problem...

**Comment by elek** [comment]
> The fact that it worked for other chunked files means to me that this is probably a Storj bug...

Indeed, it is. I can reproduce it with simple aws cli and opened storj/gateway-mt#39

**Comment by calebcase** [comment]
> @calebcase as you wrote the original tardigrade backend, do you have any thoughts on whether we should rename it from `tardigrade` to `storj`?

Renaming would be great!

**Comment by calebcase** [comment]
I also agree generally with the notion that using S3 with the Storj Gateway is the path most users should go down first. Using Storj's native protocol effectively requires a more advanced knowledge of networking and is better for the end user in two primary situations: 1) when the user has a server with a very good network connection, cpu, and memory and needs the additional throughput, and 2) when the user's security requirements can't be met by the S3 Gateway's server side encryption. Nice work on the fact gathering and analysis!

**Comment by ncw** [comment]
I've merged the rename to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.58 - the website changes won't appear until 1.58 is released.

I think we are done with this issue now - @elek ?

**Comment by elek** [comment]
Sure, It can be closed.

Thank you very much.
